src/app/Dashboard/Charts/ChartController.tsx

```diff
@@ -0,0 +1,190 @@
+  private readonly _subscriptions: Subscription[] = [];
+
+  constructor(
+    private readonly _api: ApiService,
+    private readonly _target: TargetService,
+    private readonly _notifications: NotificationChannel,
+    private readonly _settings: SettingsService
+  ) {
+    this._subscriptions.push(
+      this._refCount$
+        .pipe(
+          map((v) => v > 0),
+          pairwise()
+        )
+        .subscribe((v) => {
+          const prev = v[0];
+          const curr = v[1];
+          if (prev && !curr) {
+            // last subscriber left
+            this._tearDown();
+          }
+          if (!prev && curr) {
+            // first subscriber joined
+            this._init();
+          }
+        })
+    );
+  }
+
+  hasActiveRecording(): Observable<boolean> {
+    return this._hasRecording$.asObservable();
+  }
+
+  attach(): Observable<number> {
+    this._refCount$.next(this._refCount$.value + 1);
+    return this._refCount$.asObservable().pipe(finalize(() => this._refCount$.next(this._refCount$.value - 1)));
+  }
+
+  requestRefresh(): void {
+    this._updates$.next();
+  }
+
+  private _hasRecording(target: Target): Observable<boolean> {
+    if (target === NO_TARGET) {
+      return of(false);
+    }
+
+    return this._api
+      .graphql<CountResponse>(
+        `
+          query ActiveRecordingsForAutomatedAnalysis($connectUrl: String) {
+            targetNodes(filter: { name: $connectUrl }) {
+              recordings {
+                active (filter: {
+                  labels: ["origin=${RECORDING_NAME}"],
+                  state: "RUNNING",
+                }) {
+                aggregate {
+                  count
+                }
+              }
+            }
+            }
+          }`,
+        { connectUrl: target.connectUrl }
+      )
+      .pipe(
+        map((resp) => {
+          const nodes = resp?.data?.targetNodes;
+          if (nodes.length === 0) {
+            return false;
+          }
+          const count = nodes[0]?.recordings?.active?.aggregate?.count;
+          return count > 0;
+        }),
+        catchError((_) => of(false)),
+        first()
+      );
+  }
+
+  private _init(): void {
+    this._subscriptions.push(
+      merge(
+        merge(
+          this._updates$.pipe(throttleTime(this._settings.chartControllerConfig().minRefresh)),
+          this._notifications.messages(NotificationCategory.ActiveRecordingCreated),
+          this._notifications.messages(NotificationCategory.ActiveRecordingDeleted),
+          this._notifications.messages(NotificationCategory.ActiveRecordingStopped)
+        ).pipe(switchMap((_) => this._target.target().pipe(first()))),
+        this._target.target()
+      )
+        .pipe(concatMap((t) => this._hasRecording(t)))
+        .subscribe((v) => {
+          this._hasRecording$.next(v);
+          if (v) {
+            this._api
+              .uploadActiveRecordingToGrafana(RECORDING_NAME)
```

Comment:
I am running into a bunch of error messages when switching between target while the metrics cards refresh. I think because internally `uploadActiveRecordingToGrafana` subscribe to `context.target.target` but the incoming value for previous target.

![Screenshot from 2023-02-09 01-41-26](https://user-images.githubusercontent.com/68053619/217737391-d0edfdd3-0d9e-4e3c-98f1-a522544a186f.png)

Reply:
Huh. I thought I had solved that. I'll try to reproduce it again.

Reply:
Are you still seeing this with the latest changes?

Reply:
Ohhh I dont see it anymore. Nicee! Not sure what happened tho.